tests/sys/ringbuffer_api/src/main.c

```diff
@@ -0,0 +1,185 @@
+
+static u32_t buffer[RINGBUFFER_SIZE];
+
+static struct {
+	u8_t length;
+	u8_t value;
+	u16_t type;
+	u32_t buffer[DATA_MAX_SIZE];
+} data[] = {
+	{ 0, 32, 1, {} },
+	{ 1, 76, 54, { 0x89ab } },
+	{ 3, 0xff, 0xffff, { 0x0f0f, 0xf0f0, 0xff00 } }
+};
+
+/*entry of contexts*/
+static void tringbuf_put(void *p)
+{
+	int index = (int)p;
+	/**TESTPOINT: ring buffer put*/
+	int ret = sys_ring_buf_put(pbuf, data[index].type, data[index].value,
+				   data[index].buffer, data[index].length);
+
+	zassert_equal(ret, 0, NULL);
+}
+
+static void tringbuf_get(void *p)
+{
+	u16_t type;
+	u8_t value, size32 = DATA_MAX_SIZE;
+	u32_t rx_data[DATA_MAX_SIZE];
+	int ret, index = (int)p;
+
+	/**TESTPOINT: ring buffer get*/
+	ret = sys_ring_buf_get(pbuf, &type, &value, rx_data, &size32);
+	zassert_equal(ret, 0, NULL);
+	zassert_equal(type, data[index].type, NULL);
+	zassert_equal(value, data[index].value, NULL);
+	zassert_equal(size32, data[index].length, NULL);
+	zassert_equal(memcmp(rx_data, data[index].buffer, size32), 0, NULL);
+}
+
+/*test cases*/
+void test_ringbuffer_init(void)
+{
+	/**TESTPOINT: init via sys_ring_buf_init*/
+	sys_ring_buf_init(&ringbuf, RINGBUFFER_SIZE, buffer);
+	zassert_true(sys_ring_buf_is_empty(&ringbuf), NULL);
+	zassert_equal(sys_ring_buf_space_get(&ringbuf), RINGBUFFER_SIZE - 1, NULL);
+}
+
+void test_ringbuffer_declare_pow2(void)
+{
+	zassert_true(sys_ring_buf_is_empty(&ringbuf_pow2), NULL);
+	zassert_equal(sys_ring_buf_space_get(&ringbuf_pow2), (1 << POW) - 1, NULL);
+}
+
+void test_ringbuffer_declare_size(void)
+{
+	zassert_true(sys_ring_buf_is_empty(&ringbuf_size), NULL);
+	zassert_equal(sys_ring_buf_space_get(&ringbuf_size), RINGBUFFER_SIZE - 1,
+		      NULL);
+}
+
+void test_ringbuffer_put_get_thread(void)
+{
+	pbuf = &ringbuf;
+	tringbuf_put((void *)0);
+	tringbuf_put((void *)1);
+	tringbuf_get((void *)0);
+	tringbuf_get((void *)1);
+	tringbuf_put((void *)2);
+	zassert_false(sys_ring_buf_is_empty(pbuf), NULL);
+	tringbuf_get((void *)2);
+	zassert_true(sys_ring_buf_is_empty(pbuf), NULL);
+}
+
+void test_ringbuffer_put_get_isr(void)
+{
+	pbuf = &ringbuf;
+	irq_offload(tringbuf_put, (void *)0);
+	irq_offload(tringbuf_put, (void *)1);
+	irq_offload(tringbuf_get, (void *)0);
+	irq_offload(tringbuf_get, (void *)1);
+	irq_offload(tringbuf_put, (void *)2);
+	zassert_false(sys_ring_buf_is_empty(pbuf), NULL);
+	irq_offload(tringbuf_get, (void *)2);
+	zassert_true(sys_ring_buf_is_empty(pbuf), NULL);
+}
+
+void test_ringbuffer_put_get_thread_isr(void)
+{
+	pbuf = &ringbuf;
+	tringbuf_put((void *)0);
+	irq_offload(tringbuf_put, (void *)1);
+	tringbuf_get((void *)0);
+	irq_offload(tringbuf_get, (void *)1);
+	tringbuf_put((void *)2);
+	irq_offload(tringbuf_get, (void *)2);
+}
+
+void test_ringbuffer_pow2_put_get_thread_isr(void)
+{
+	pbuf = &ringbuf_pow2;
+	tringbuf_put((void *)0);
+	irq_offload(tringbuf_put, (void *)1);
+	tringbuf_get((void *)0);
+	irq_offload(tringbuf_get, (void *)1);
+	tringbuf_put((void *)1);
```

Comment:
Might be a good idea to avoid repetition here by refactoring the test.

For instance,  `test_ringbuffer_put_get_thread_isr()` and `test_ringbuffer_size_put_get_thread_isr()` are similar tests, and they're both calling this with `(void *)2`.

Reply:
yes, this test has lots of repetition, but it is a test, as long as it is functionally correct we should be fine. I just took this as is from QA and do not think it is worth the trouble of refactoring.

Reply:
I'm fine with the tests not being refactored, but it would be nice to understand at least why the parameters to put/get are different in these tests.